stdlib/pqueue.mli

```diff
@@ -0,0 +1,299 @@
+        elements of list [l]. Runs in linear time. *)
+
+    val of_seq: elt Seq.t -> t
+    (** [of_seq s] returns a new priority queue containing the
+        elements of sequence [s]. Runs in linear time. *)
+
+    (** {1:iteration Iteration}
+
+        The order in which the elements of a priority queue are
+        traversed is unspecified.
+
+        It is a programming error to mutate a priority queue (by
+        adding or removing elements) during an iteration of the queue.
+        Such an error may be detected and signaled by the backing dynamic
+        array implementation, but this is not guaranteed. *)
+
+    val iter: (elt -> unit) -> t -> unit
+    (** [iter f q] calls [f] on each element of [q] in some
+        unspecified order. *)
+
+    val fold: ('acc -> elt -> 'acc) -> 'acc -> t -> 'acc
+    (** [fold f accu q] is [(f (... (f (f accu x1) x2) ...) xn)]
+        where [x1,x2,...,xn] are the elements of [q] in some
+        unspecified order. *)
+
+    val to_seq: t -> elt Seq.t
+    (** [to_seq q] is the sequence of elements of [q] in some
+        unspecified order. *)
+
+  end
+(** Output signature of the functor {!MakeMin}. *)
+
+module MakeMin(E: OrderedType) : Min with type elt := E.t
+(** Functor building an implementation of the min-priority queue
+    structure given a totally ordered type for elements. *)
+
+module type Max =
+  sig
+    type t
+    type elt
+    val create: unit -> t
+    val length: t -> int
+    val is_empty: t -> bool
+    val add: t -> elt -> unit
+    val add_seq: t -> elt Seq.t -> unit
+    exception Empty
+    val max_elt: t -> elt
+    val max_elt_opt: t -> elt option
+    val pop_max: t -> elt
+    val pop_max_opt: t -> elt option
+    val remove_max: t -> unit
+    val clear: t -> unit
+    val copy: t -> t
+    val of_array: elt array -> t
+    val of_list: elt list -> t
+    val of_seq: elt Seq.t -> t
+    val iter: (elt -> unit) -> t -> unit
+    val fold: ('acc -> elt -> 'acc) -> 'acc -> t -> 'acc
+    val to_seq: t -> elt Seq.t
+end
+(** Output signature of the functor {!MakeMax}. *)
+
+module MakeMax(E: OrderedType) : Max with type elt := E.t
+(** Functor building an implementation of the max-priority queue
+    structure given a totally ordered type for elements. *)
+
+(** {1 Polymorphic priority queues}
+
+    The following, more complex functors create polymorphic queues of
+    type ['a t], just like other polymorphic containers (lists,
+    arrays...). They require a notion of "polymorphic elements" ['a
+    elt] that can be compared without depending on choice of ['a].
+
+    One usage scenario comes from certain polymorphic containers that
+    embed priority information, for example:
+    {[
+      type 'a task = {
+        run : unit -> 'a;
+        cancel : unit -> unit;
+        priority : int;
+      }
+
+      module TaskQueue = PQueue.MakeMaxPoly(struct
+        type 'a elt = task
+        let compare t1 t2 = Int.compare t1.priority t2.priority
+      end)
+    ]}
```

Comment:
Could we perhaps try to find a good example for that ? Or failing to simply drop this example: this `'a task` is [totally unconvincing](https://gist.github.com/dbuenzli/22c177c7e4daaae389b1013f5344acd3?permalink_comment_id=5053464#gistcomment-5053464) to me. I think the separate priorities that follows is sufficient to justify the poly signatures.